packages/ember-routing/lib/system/router.js

```diff
@@ -694,6 +704,33 @@ var EmberRouter = EmberObject.extend(Evented, {
     };
   },
 
+  _normalizeQueryParams(leafRouteName, contexts, queryParams) {
+    var state = calculatePostTransitionState(this, leafRouteName, contexts);
+    var handlerInfos = state.handlerInfos;
+    var appCache = this._bucketCache;
+    stashParamNames(this, handlerInfos);
+
+    for (var i = 0, len = handlerInfos.length; i < len; ++i) {
+      var route = handlerInfos[i].handler;
+      var qpMeta = get(route, '_qp');
+
+      for (var j = 0, qpLen = qpMeta.qps.length; j < qpLen; ++j) {
+        var qp = qpMeta.qps[j];
+
+        var presentProp = qp.prop in queryParams  && qp.prop ||
```

Comment:
Can you please break this down and add more local vars to make this more readable